### PR TITLE
Add validation for dimension values

### LINKED ISSUE #68: Runtime Validations
allowed:

`putDimensions({ Method: "GET", StatusCode: "200" })`

not allowed:

`putDimensions({ Method: "GET", StatusCode: 200 })`

### PR DESCRIPTION
**Pull request by markkuhn**
*Issue #, if available:* #68

Related issues:
* https://github.com/awslabs/aws-embedded-metrics-python/issues/88
* https://github.com/awslabs/aws-embedded-metrics-java/issues/112
* https://github.com/awslabs/aws-embedded-metrics-dotnet/issues/35

*Description of changes:*
Add function to strip non-ascii characters and stringify numbers for dimension values.

[UPDATE] Instead of silently dropping non-compliant characters, the client will now throw an error if an invalid dimension is found according to specifications [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_Dimension.html) (breaking change).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by gordonpn — APPROVED** [review]
Looks good to me!

**Review by Himtanaya — APPROVED** [review]

File: src/serializers/LogSerializer.ts [hunk 1]
```diff
@@ -38,21 +38,21 @@ export class LogSerializer implements ISerializer {
...
-        `Account for default dimensions if not using set_dimensions.`;
-        throw new DimensionSetExceededError(errMsg)
+        const errMsg =
+          `Maximum number of dimensions allowed are ${Constants.MAX_DIMENSION_SET_SIZE}.` +
+          `Account for default dimensions if not using set_dimensions.`;
+        throw new DimensionSetExceededError(errMsg);
       }
+
+      // Stringify and remove non-ascii characeters (allow only 0x20-0x7E)
+      keys.forEach(key => (dimensionSet[key] = String(dimensionSet[key]).replace(/[^\x20-\x7F]/g, '')));
```
Comment (giancarlokc) [inline src/serializers/LogSerializer.ts:52]: This is a bit in the other direction from the current pattern of validation for dimension set size. In this case we're silently removing the characters, while for dimensions we are throwing an exception.

I think the same principle of falling on the side of failing fast and informing the user that a record cannot be processed is better than silently fixing things. Let me know what you think.

Comment (markkuhn) [reply src/serializers/LogSerializer.ts:52]: Spoke privately. Editing customer logs is not the best way to resolve this issue. Will work on throwing an error if invalid characters are used.

File: src/logger/MetricsContext.ts [hunk 2]
```diff
@@ -107,14 +108,46 @@ export class MetricsContext {
...
+        throw new InvalidDimensionError(`Dimension value ${value} is invalid`);
+      }
+
+      if (
+        !MetricsContext.isAscii(value) ||
+        value.trim().length == 0 ||
+        value.length > Constants.MAX_DIMENSION_VALUE_LENGTH
+      ) {
+        throw new InvalidDimensionError(`Dimension value ${value} is invalid`);
+      }
```
Comment (gordonpn) [inline src/logger/MetricsContext.ts:141]: I'm wondering if the error should be more descriptive?

E.g., if the user has a key that is more than 250 characters long with one or more non-ascii characters as well, they would receive the same error message, wouldn't they?

Comment (markkuhn) [reply src/logger/MetricsContext.ts:141]: What do you suggest I change it to?

File: src/logger/MetricsContext.ts [hunk 3]
```diff
@@ -107,14 +108,46 @@ export class MetricsContext {
...
+    });
+  }
+
+  /**
+   * Check if the string contains only ASCII characters.
+   * @param str string to check
+   */
+  public static isAscii(str: string): boolean {
+    return /^[\x20-\x7F]*$/.test(str);
```
Comment (giancarlokc) [inline src/logger/MetricsContext.ts:150]: Out of curiosity, have we checked if there's a native Node.js way for checking this? Or perhaps a widely used library that can do this check for us?

File: src/logger/MetricsContext.ts (hunk 2 above)
Comment (gordonpn) [reply src/logger/MetricsContext.ts:141]: Just a more specific error message should be fine.

Similar to what GC mentioned here: https://github.com/awslabs/aws-embedded-metrics-node/pull/131#discussion_r950219522

File: src/logger/MetricsContext.ts (hunk 3 above)
Comment (markkuhn) [reply src/logger/MetricsContext.ts:150]: I could not find a native Node.js way or any widely used library that can do this for us.

Comment (johnjang) [reply src/logger/MetricsContext.ts:150]: Not much of a better suggestion, but I believe latest JS (2018+) has this regex that specifically checks for ASCII characters: ```/[\p{ASCII}]+/u``` ([test](https://regex101.com/r/gGjNmr/1)). Not sure if its applicable here. 
Personally, I don't think using regex is a bad option though as it is relatively simple enough and future proof (js string is always in UTF-16 I believe?)

Comment (markkuhn) [reply src/logger/MetricsContext.ts:150]: I agree that it is relatively future proof, but we could always use [validator](https://www.npmjs.com/package/validator) to check for ASCII.

**Review by gordonpn — APPROVED** [review]

**Review by giancarlokc — APPROVED** [review]

**Pull request merged**